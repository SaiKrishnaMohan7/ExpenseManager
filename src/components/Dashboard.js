import React from 'react';

const Dashboard = () => {
    return (
    <div>
        this i from the dashboard component
    </div>        

    );
};

export default Dashboard;