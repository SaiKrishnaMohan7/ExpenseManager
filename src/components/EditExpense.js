import React from 'react';

const EditExpense = (props) => {
    // url is dynamic and the dynamic can be accessed via props.match.params and key-value obj
    return (
        <p>Editing Expense {props.match.params.id}</p>
    );
};

export default EditExpense;