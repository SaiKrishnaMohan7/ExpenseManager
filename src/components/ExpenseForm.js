import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component{
  constructor(props){
    super(props);

    this.state = {
      amount: props.expense ? props.expense.amount.toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      description: props.expense ? props.expense.description : '',
      focused: false,
      error: '',
      note: props.expense ? props.expense.note : ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;

    this.setState(() => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;

    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState({
        createdAt: createdAt
      });
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState({ focused: focused });
  };

  onSubmit = (e) => {
    const { amount, description, createdAt, note } = this.state;
    e.preventDefault();
    if (!(amount && description)) {
      this.setState({ error: 'Both amount and description are required' });
    } else {
      const { onSubmit } = this.props;

      this.setState({ error: '' });
      onSubmit({
        amount: parseFloat(amount, 10),
        createdAt: createdAt.valueOf(),
        description,
        note 
      });
    }
  }

  render () {
    const {
      amount,
      createdAt,
      description,
      error,
      focused,
      note
    }  = this.state;
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={description}
            onChange={this.onDescriptionChange}
          />
          
          <input
            type="text"
            placeholder="Amount"
            value={amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={createdAt}
            onDateChange={this.onDateChange}
            focused={focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => (false)}
          />

          <textarea 
            placeholder="Add Some notes (optional)"
            value={note}
            onChange={this.onNoteChange}
          />

          <button>Add Expense</button>
        </form>
        {error && (<p>{error}</p>)}
      </div>
    );
  }
}