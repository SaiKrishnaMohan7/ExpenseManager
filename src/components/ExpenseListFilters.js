import React, { Component } from 'react';
import {connect} from 'react-redux';
import { DateRangePicker } from 'react-dates';

import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filtersActionGen';


export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState({ calendarFocused });
  };

  onSortChange = (e) => {
    const { sortByAmount, sortByDate } = this.props;
    const selected = e.target.value;
    if (selected === 'date')
      sortByDate();
    if(selected === 'amount')
      sortByAmount();
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  render(){
    const { filters } = this.props;
    const { text, sortBy, startDate, endDate } = filters;
    const { calendarFocused } = this.state;

    return (
      <div>
        <input type="text" value={text} onChange={this.onTextChange}/>
        <select 
          value={sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={startDate}
          endDate={endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
          displayFormat="DD/MM/YYYY"
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);