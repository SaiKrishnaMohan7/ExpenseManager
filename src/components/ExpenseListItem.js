import React, {Fragment} from 'react';
import { connect } from 'react-redux';

import {removeExpense} from '../actions/expensesActionGen';

const ExpenseListItem = ({id, description, amount, createdAt, dispatch}) => {
  return (
    <Fragment>
      <h3>{description}</h3>
      <p>{amount} - {createdAt}</p>
      <button onClick={() => {dispatch(removeExpense(id));}}>
          Remove
      </button>
    </Fragment>
  );
};

// using connect() without args passes dispatch as a prop to the connected component
export default connect()(ExpenseListItem);