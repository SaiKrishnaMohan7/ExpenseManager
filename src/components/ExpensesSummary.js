import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import getVisibleExpense from '../selectors/expensesSelector';
import totalExpense from '../selectors/totalExpense';

export const ExpensesSummary = ({expenseCount, expenseTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const expenseTotalFormatted = numeral(expenseTotal / 100).format('$0,0.00');
  return (
    <Fragment>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {expenseTotalFormatted}
      </h1>
    </Fragment>
  );
}

const mapStateToProps = state => {
  const visibleExpenses = getVisibleExpense(state.expenses.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: totalExpense(visibleExpenses)
  }
};

export default connect(mapStateToProps)(ExpensesSummary);
