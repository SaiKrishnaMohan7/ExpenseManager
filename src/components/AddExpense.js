import React, { Component } from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expensesActionGen';

export class AddExpense extends Component {
	onSubmit = (newExpense) => {
		this.props.addExpense(addExpense(newExpense));
		// Route to dashboard page after item is added
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
					<h1>Add Expense</h1>
					<ExpenseForm
						onSubmit={this.onSubmit}
					/>
			</div>        
		);
	}
}

const mapDispatchToProps = (dispatch) => {
	return {
		addExpense: (expense) => {
			dispatch(addExpense(expense));
		}
	};
};

export default connect(undefined, mapDispatchToProps)(AddExpense);