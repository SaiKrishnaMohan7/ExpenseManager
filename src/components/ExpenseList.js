import React from 'react';
import {connect} from 'react-redux';

import ExpenseListItem from './ExpenseListItem'
import expensesSelector from './../selectors/expensesSelector';

export const ExpenseList = (props) => {
  return (
    <div>
      <h2>Expense List</h2>
        {
          props.expenses.map((expense) => {
            // By using spread operator here, the component will have access ot properties on props as opposed to props.expense if
            // we did this: expense={expense}
            return (
              <ExpenseListItem key={expense.id} {...expense } />
            );
          })
        }
    </div>
  );
};

const mapStateToProps = (state) => {
  // To make sure filters get applied when user selects a filter
  return {
    expenses: expensesSelector(state.expenses, state.filters)
  };
};

// connect() returns a fucntion so it has to be invoked!
// Component will re-render as state chnages, no neeed to subscribe, maybe its implicit
export default connect(mapStateToProps)(ExpenseList);