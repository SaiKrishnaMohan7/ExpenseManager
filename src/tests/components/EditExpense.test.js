import React from 'react';
import { shallow } from 'enzyme';

import { EditExpense } from '../../components/EditExpense';
import { expenses } from '../testData/testData';


let props, wrapper;
describe('Edit Expense Component', () => {
  beforeEach(() => {
    props = {
      editExpense: jest.fn(),
      removeExpense: jest.fn(),
      history: { push: jest.fn() },
      expense: expenses[0],
    };
    wrapper = shallow(<EditExpense {...props} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle submission ', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')({...expenses[0], note: 'test'});

    expect(props.editExpense).toHaveBeenCalled();
    expect(props.history.push).toHaveBeenLastCalledWith('/');
  });

  it('should handle removing an expense', () => {
    wrapper.find('button').simulate('click');

    expect(props.removeExpense).toHaveBeenCalled();
    expect(props.history.push).toHaveBeenLastCalledWith('/');
  });
});