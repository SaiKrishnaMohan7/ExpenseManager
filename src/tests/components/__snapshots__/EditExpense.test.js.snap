// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Expense Component should match snapshot 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 20000,
        "createdAt": 0,
        "description": "car",
        "id": "1",
        "note": "bought a car",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
